src/contracts/deposits-registry/deposits-registry.service.ts

```diff
@@ -0,0 +1,223 @@
+
+@Injectable()
+export class DepositRegistryService {
+  constructor(
+    @Inject(WINSTON_MODULE_NEST_PROVIDER) private logger: LoggerService,
+    private providerService: ProviderService,
+    private repositoryService: RepositoryService,
+
+    private sanityChecker: DepositRegistrySanityCheckerService,
+    private fetcher: DepositsRegistryFetcherService,
+    private store: DepositsRegistryStoreService,
+  ) {}
+
+  public async handleNewBlock(blockNumber: number): Promise<void> {
+    if (blockNumber % DEPOSIT_EVENTS_CACHE_UPDATE_BLOCK_RATE !== 0) return;
+
+    // The event cache is stored with an N block lag to avoid caching data from uncle blocks
+    // so we don't worry about blockHash here
+    await this.updateEventsCache();
+  }
+
+  public async initialize() {
+    await this.store.initialize();
+    const cachedEvents = await this.store.getEventsCache();
+    await this.sanityChecker.initialize(cachedEvents);
+    // it is necessary to load fresh events before integrity check
+    // because we can only compare roots of the last 128 blocks.
+    await this.updateEventsCache();
+  }
+
+  /**
+   * Gets node operators data from cache
+   * @returns event group
+   */
+  public async getCachedEvents(): Promise<VerifiedDepositEventsCache> {
+    const { headers, ...rest } = await this.store.getEventsCache();
+    const deploymentBlock = await this.fetcher.getDeploymentBlockByNetwork();
+
+    return {
+      headers: {
+        ...headers,
+        startBlock: Math.max(headers.startBlock, deploymentBlock),
+        endBlock: Math.max(headers.endBlock, deploymentBlock),
+      },
+      ...rest,
+    };
+  }
+
+  /**
+   * Updates the cache deposited events
+   * The last N blocks are not stored, in order to avoid storing reorganized blocks
+   */
+  public async updateEventsCache(): Promise<void> {
+    const fetchTimeStart = performance.now();
+
+    const [currentBlock, initialCache] = await Promise.all([
+      this.providerService.getBlock(),
+      this.getCachedEvents(),
+    ]);
+
+    const { number: currentBlockNumber, hash: currentBlockHash } = currentBlock;
+    const firstNotCachedBlock = initialCache.headers.endBlock + 1;
+    const toBlock = currentBlockNumber - DEPOSIT_EVENTS_CACHE_LAG_BLOCKS;
+
+    const totalEventsCount = initialCache.data.length;
+    let newEventsCount = 0;
+
+    // verify blockchain
+    const isCacheValid = this.sanityChecker.verifyCacheBlock(
+      initialCache,
+      currentBlockNumber,
+    );
+
+    if (!isCacheValid) return;
+
+    for (
+      let block = firstNotCachedBlock;
+      block <= toBlock;
+      block += DEPOSIT_EVENTS_STEP
+    ) {
+      const chunkStartBlock = block;
+      const chunkToBlock = Math.min(toBlock, block + DEPOSIT_EVENTS_STEP - 1);
+
+      const chunkEventGroup = await this.fetcher.fetchEventsFallOver(
+        chunkStartBlock,
+        chunkToBlock,
+      );
+
+      await this.sanityChecker.verifyEventsChunk(
```

Comment:
I think better don't add `chunkStartBlock`, `chunkToBlock` as parameters. just move long from verifyEventsChunk and remove chunk from name